### PR TITLE
wip: state refactor

### PR DESCRIPTION
**Pull request by davidrusu**

### CODE REVIEW

File: src/mvba/vcbc/message.rs [hunk 1]
```diff
@@ -1,8 +1,7 @@
 use crate::mvba::proposal::Proposal;
 
-
 #[derive(Debug, serde::Serialize, serde::Deserialize)]
-pub(crate) enum Message {
```
Comment (b00f) [inline src/mvba/vcbc/message.rs:5]: I believe we should not expose the internal format of messages.

File: src/mvba/vcbc/mod.rs [hunk 2]
```diff
@@ -19,12 +13,19 @@ use std::rc::Rc;
...
+    Echo,
+    Deliver,
+}
+
 // VCBC is a verifiably authenticatedly c-broadcast protocol.
 // Each party $P_i$ c-broadcasts the value that it proposes to all other parties
 // using verifiable authenticated consistent broadcast.
 pub(crate) struct Vcbc {
     ctx: context::Context,
-    state: Box<dyn State>,
```
Comment (b00f) [inline src/mvba/vcbc/mod.rs:27]: There are two way to implement state design pattern in Rust, using `Box<dyn state>` or `enum`. I prefer the first, If you don't mind let's do in this way. The actual algorithm is the same.

File: src/mvba/vcbc/mod.rs [hunk 3]
```diff
@@ -45,31 +46,93 @@ impl Vcbc {
...
         Self {
             ctx,
-            state: Box::new(ProposeState),
+            state: State::Propose,
         }
     }
 
     // propose sets the proposal and broadcast propose message.
     pub fn propose(&mut self, proposal: &Proposal) -> Result<()> {
-        debug_assert_eq!(proposal.proposer_id, self.ctx.proposer_id);
```
Comment (b00f) [inline src/mvba/vcbc/mod.rs:54]: I am not sure why you removed this debug assert. It's not bad to have it here.

File: src/mvba/vcbc/mod.rs [hunk 4]
```diff
@@ -1,16 +1,10 @@
 pub(super) mod context;
 pub(super) mod message;
-pub(super) mod state;
 
-mod deliver;
```
Comment (b00f) [inline src/mvba/vcbc/mod.rs:5]: I can see you deleted the state files. I prefer to not. Let's break the code and make logic separate per state.

File: src/mvba/vcbc/mod.rs (hunk 2 above)
Comment (davidrusu) [reply src/mvba/vcbc/mod.rs:27]: We don't use traits in the code base when we can avoid it. In this case we can avoid it so please use them. It'll make the implementation fit into the rest of the codebase.

e.g. it took me some time to figure out what was happening with the trait based State, whereas this enum style will be familiar to anyone on the team.

File: src/mvba/vcbc/mod.rs (hunk 3 above)
Comment (davidrusu) [reply src/mvba/vcbc/mod.rs:54]: It was being check for below: https://github.com/maidsafe/sn_consensus/pull/70/files/8cc9e360b8c0ca75e655a7036a5b5561648ccaf1#diff-af91cddd2adf4b2869052ddc812b17294246da37b01643cdc33edfc550f7f04cR60

But yeah I could go either way, I don't mind keeping it

File: src/mvba/vcbc/message.rs (hunk 1 above)
Comment (b00f) [reply src/mvba/vcbc/message.rs:5]: Agree. Let's let caller deal with serialization and de-serialization.

**Pull request closed without merging**